doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,246 @@
+
+The informationn of PHY device using clause 22 mdio can also be added to the gearbox\_config.json file. It will help to decide the mdio access mode at runtime.
+
+![syncd gbsyncd MDIO IPC](images/MDIO-IPC.png)
+
+The VendorSai class constructor is overloaded to handle the dynamic linking of the libraries. The  syncd class VendorSai currently only has a constructor with no argument. The linked SAI library path is determined at syncd build time. New constructors with arguments for the VendorSai class is introduced as the overloading constructors. The new constructors can either have the SAI/PAI library path as the argument or have PHY id as one of the arguments. When the constructor with PHY id as the argument is used, the syncd can index into the "phys" section of gearbox configuration file "gearbox\_config.json" to get the PAI library path and MDIO access library path at run time.
+
+### High-Level Design 
+
+The high level design of the gbsyncd mdio access function using the mdio bus from the switch NPU and using single gbsyncd docker in SONiC image can be covered in the following points:
+		
+	- It is an enhancement to the sonic gearbox framework.
+	- Syncd, SAI and some sonic device files are modified.
+	- There are 3 repositories that would be changed, sonic-sairedis, SAI and sonic-buildimage.
+	- gbsyncd will have dependencies on syncd to create the Unix IPC socket. 
+	- Syncd will have new classes for the MDIO IPC server and for the parser of gearbox configuration.
+	- There is a new MDIO IPC client library is added.
+	- External PHY firmware download time always has performance requirements/impact.
+	- For debugging, beside the syslog logging, the Unix socket IPC mechanism can be simulated by socat.
+	- The current change is only limited to the Broadcom switch NPU platform.
+	- SAI switch api has added the mido clause 22 access functions.
+	- The SaiInterface class has added the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- The VendoSai class overrides the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- New keys are added to the gearbox configuration file "gearbox\_config.json".
+
+When the VendorSai class constructor with PHY id argument is used in syncd.cpp, we can assume a simple case where syncd context number is the same as the PHY id.
+
+	int syncd_main(int argc, char **argv)
+	{
+	    ...
+	    if (commandLineOptions->m_globalContext != 0) {   
+	        auto vendorSai = std::make_shared<VendorSai>(
+	        (int)commandLineOptions->m_globalContext,
+	        (int)commandLineOptions->m_globalContext);
+	        ...
+	    } else {   
+	        auto vendorSai = std::make_shared<VendorSai>();
+	        ...
+	    }
+	    ...
+	}
+
+For the case that multiple PHY ids aggregate in a single syncd context, it is a new separate topic. We don't discuss the topic in this HLD.
```

Comment:
Currently we are in the case of "multiple PHY ids aggregate in a single syncd context" by default. Need consider it for backwards compatibility. Even different cases could coexist.

Reply:
This design does not block other software components/changes to manage multiple PHYs in a single context.

Reply:
If `commandLineOptions->m_globalContext != 0`, will it only create one vendorSai? In the design, looks a vendorSai will dynamically load a PAI library for one PHY. How to load multiple PAI libs?

Reply:
This design does not cover the PHY aggregation issue. When there is no PHY aggregation, each PHY will have its own context and each PHY is controlled by one gbsyncd process. Each gbsyncd process will have a different context.

Reply:
Currently Broncos/Broadcom PAI for PikeZ is working with PHY aggregation, i.e. one gbsyncd process with one context manages multiple PHYs. This design should cover it.

Reply:
Please point us the PR of Broncos/Broadcom PAI for PikeZ PHY aggregation.

Reply:
Please refer to PR https://github.com/sonic-net/sonic-buildimage/pull/11362, file device/arista/x86_64-arista_720dt_48s/Arista-720DT-48S/context_config.json.

Reply:
I only see that context of a json file but I don't see the code processing the context. In order for us to adapt our design for the phy aggregation, we need to know how the context_config.json is processed and used.
We only can design based on what we know. Our original design has listed context limitation in the HLD. We can have a follow up HLD to overcome the limitation in this HLD when we see the design of the PHY aggregation.

Reply:
im not sure about idea passing globalContext to VendorSai since global context is concept of syncd and sairedis not VendorSai

Reply:
We may add another command line options to specify PHY id. I will double check if global context is needed in VendorSai. Our existing code use VendorPaiDll class which inherits from VendorSai class. We found that we need to override most member functions but with a small change in each member function. We come to the conclusion that change VendorSai itself makes more sense.